### PR TITLE
[storage] expect_value_by_version does not seek

### PR DESCRIPTION
**Pull request by msmouse**
### Description

It's faster, although not sigificantly

```
msmouse@ubuntu-dev:~/work/ait3-backup$ ./db-backup  one-shot backup  --max-chunk-size 10000000 state-snapshot --state-snapshot-epoch 118 local-fs --dir local
2022-09-05T17:43:01.534854Z [main] INFO crates/aptos-push-metrics/src/lib.rs:71 PUSH_METRICS_ENDPOINT env var is not set. Skipping sending metrics.
2022-09-05T17:43:01.541599Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:63 State snapshot backup started, for epoch 118.
2022-09-05T17:43:05.467167Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":41068,"values_per_second":13689}
2022-09-05T17:43:09.371326Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":82135,"values_per_second":11733}
2022-09-05T17:43:13.283234Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":123200,"values_per_second":11200}
2022-09-05T17:43:17.210570Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":164277,"values_per_second":10951}
2022-09-05T17:43:21.090090Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":205341,"values_per_second":10807}
2022-09-05T17:43:25.041937Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":246402,"values_per_second":10713}
2022-09-05T17:43:28.972751Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":287468,"values_per_second":10647}
2022-09-05T17:43:32.937181Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":328539,"values_per_second":10598}
2022-09-05T17:43:36.817166Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":369600,"values_per_second":10560}
2022-09-05T17:43:40.767353Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":410662,"values_per_second":10529}
^C
msmouse@ubuntu-dev:~/work/ait3-backup$ ./db-backup-no-expect  one-shot backup  --max-chunk-size 10000000 state-snapshot --state-snapshot-epoch 118 local-fs --dir local
2022-09-05T17:43:47.687852Z [main] INFO crates/aptos-push-metrics/src/lib.rs:71 PUSH_METRICS_ENDPOINT env var is not set. Skipping sending metrics.
2022-09-05T17:43:47.695658Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:63 State snapshot backup started, for epoch 118.
2022-09-05T17:43:51.832369Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":41068,"values_per_second":10267}
2022-09-05T17:43:56.071143Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":82135,"values_per_second":10267}
2022-09-05T17:44:00.221019Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":123200,"values_per_second":10266}
2022-09-05T17:44:04.387256Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":164277,"values_per_second":10267}
2022-09-05T17:44:08.553962Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":205341,"values_per_second":10267}
2022-09-05T17:44:12.722405Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":246402,"values_per_second":9856}
2022-09-05T17:44:16.926940Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":287468,"values_per_second":9912}
2022-09-05T17:44:21.124840Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":328539,"values_per_second":9955}
2022-09-05T17:44:25.317103Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":369600,"values_per_second":9989}
2022-09-05T17:44:29.660370Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":410662,"values_per_second":10016}
2022-09-05T17:44:33.836584Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":451731,"values_per_second":9820}
2022-09-05T17:44:38.085339Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":492807,"values_per_second":9856}
2022-09-05T17:44:42.298589Z [main] INFO storage/backup/backup-cli/src/backup_types/state_snapshot/backup.rs:110 Chunk written. {"last_idx":533877,"values_per_second":9886}
```

### Test Plan

existing tests and manually backup state snapshot

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3867)

### CODE REVIEW

File: storage/aptosdb/src/state_store/mod.rs [hunk 1]
```diff
@@ -251,15 +252,21 @@ impl StateDb {
...
+            .ok_or_else(|| {
+                anyhow!(
+                    "State Value is missing for key {:?} by version {}",
+                    state_key,
+                    version
+                )
+            })?
+            .ok_or_else(|| {
+                anyhow!(
+                    "State Value is missing for key {:?} by version {}, found tombstone.",
```
Comment (lightmark) [inline storage/aptosdb/src/state_store/mod.rs:266]: Maybe not a tombstone...

**Review by lightmark — APPROVED** [review]

**Review by JoshLind — APPROVED** [review]
LGTM. Not sure why it's slightly faster though 🤣

**Comment by msmouse** [comment]
it doesn't work since we decided not to match the versions recorded on JMT with the actual creation version (but rather the snapshot version)

**Comment by msmouse** [comment]
wrt making it concurrent -- we sorta think it's nice for it to be slow -- to not affect VM performance, flushing things out of cache too quickly, etc.

@grao1991 brought up a great idea that we probably can limit ranges of keys we serve according to, for example, the nodes peer_id -- this way cache on each node is better utilized. @JoshLind

**Pull request closed without merging**